File: plugins/mnemuron/scripts/sync-protocol.mjs
// Packaged verbatim with the OpenClaw adapter; parity is verified by the contract test.
import { createHash, randomUUID } from 'node:crypto';
import { existsSync, mkdirSync, readFileSync, readdirSync, renameSync, rmdirSync, statSync, unlinkSync, writeFileSync, openSync, fsyncSync, closeSync, linkSync } from 'node:fs';
import path from 'node:path';

export const RESPONSE_LIMIT = 2 * 1024 * 1024;
const hash = value => createHash('sha256').update(value).digest('hex');
const states = new Set(['pending','retry_wait','paused_auth','quarantined','blocked_protocol','blocked_reconciliation','blocked_gap']);
export const protocolError = (code, statusCode) => Object.assign(new Error(`Mnemuron transport: ${code}${statusCode ? ` (${statusCode})` : ''}.`), {errorCode:code,statusCode});
export function decodeResponse(status, text, retryAfter=null) {
  let data;
  try { data=JSON.parse(text || '{}'); } catch {}
  const code=typeof data?.error_code==='string' && /^[A-Z][A-Z0-9_]{0,63}$/.test(data.error_code) ? data.error_code : null;
  if(status<200 || status>=300) throw Object.assign(protocolError(code || (status>=300&&status<400?'REDIRECT_BLOCKED':`HTTP_${status}`),status),{responseData:data,retryAfter});
  if (!data || typeof data!=='object' || Array.isArray(data)) throw protocolError('INVALID_RESPONSE_JSON',status);
  return data;
}
export function retryState(error, previous={}, {now=Date.now(), random=Math.random, credential=''}={}) {
  const status=error.statusCode ?? null, code=error.errorCode || error.responseData?.error_code || error.code || 'NETWORK_ERROR';
  const attempt=(previous.attempt_count || 0)+1;
  const result={...previous,attempt_count:attempt,last_error_code:/^[A-Z0-9_]{1,64}$/.test(code)?code:'NETWORK_ERROR',last_http_status:status,
    first_failed_at:previous.first_failed_at || new Date(now).toISOString(),next_retry_at:null,credential_fingerprint:credential};
  if(status===401 || status===403) result.state='paused_auth';
  else if(code==='RECEIPT_MISMATCH' || status===409) result.state='blocked_reconciliation';
  else if(status===413 || ([400,422].includes(status) && ['INVALID_PAYLOAD','CONTENT_TOO_LONG','SCOPE_MISMATCH','INVALID_MEMORY_SCOPE','REQUEST_BODY_TOO_LARGE'].includes(code))) result.state='quarantined';
  else if(code==='RESPONSE_TOO_LARGE' || code==='INVALID_RESPONSE_JSON' || code==='REDIRECT_BLOCKED' || (status && status<500 && status!==429)) result.state='blocked_protocol';
  else {
    result.state='retry_wait';
    let delay=Math.min(300000,1000*2**Math.min(attempt-1,18))*(0.5+random()*0.5);
    if(status===429 && error.retryAfter) {
      const value=String(error.retryAfter).trim();
      const seconds=/^\d+(?:\.\d+)?$/.test(value)?Number(value)*1000:NaN;
      const requested=Number.isFinite(seconds)?now+seconds:Date.parse(value);
      if(Number.isFinite(requested) && requested>now && requested<=8.64e15) delay=Math.max(delay,requested-now);
    }
    result.next_retry_at=new Date(now+delay).toISOString();
  }
  return result;
}
export function validateAcceptance(kind, item, result) {
  const payload=item.payload;
  if(kind==='event') {
    const id=payload?.event?.event_id;
    if(typeof id==='string' && id.length>0 && result?.status==='accepted' && result.received===1
      && [result.inserted,result.duplicate].every(n=>Number.isInteger(n) && n>=0 && n<=1) && result.inserted+result.duplicate===1
      && Array.isArray(result.accepted_event_ids) && result.accepted_event_ids.length===1 && result.accepted_event_ids[0]===id) return;
  } else {
    const receipt=kind==='receipt', delivery=result?.delivery;
    const eventKey=receipt?'receipt_event_id':'event_id', attemptKey=receipt?'receipt_id':'attempt_id';
    const records=receipt?delivery?.receipts:delivery?.attempts;
    const attempt=records?.find(record=>record[attemptKey]===payload?.[attemptKey]);
    const ids=receipt?attempt?.receipt_event_ids:attempt?.event_ids;
    const phaseField=payload?.phase==='acknowledged'?'acknowledged_at':payload?.phase==='failed'?'failed_at':receipt?'delivered_at':'injected_at';
    if(result?.[eventKey]===payload?.[eventKey] && result.inserted+result.duplicate===1
      && delivery?.resume_id===item.resume_id && delivery?.preview_version===payload.preview_version
      && attempt?.session_id===payload.session_id && attempt?.workstream_id===payload.workstream_id
      && (payload.phase==='delivered' || attempt?.turn_id===payload.turn_id)
      && attempt?.[phaseField]===payload.occurred_at && ids?.includes(payload[eventKey])
      && (payload.phase!=='acknowledged' || attempt.ack_complete===true)) return;
  }
  throw protocolError('RECEIPT_MISMATCH');
}
export function atomicState(file, value) {
  const temporary=`${file}.${process.pid}.${randomUUID()}.tmp`;
  const fd=openSync(temporary,'wx',0o600);
  try {writeFileSync(fd,JSON.stringify(value)+'\n');fsyncSync(fd);}finally{closeSync(fd);}
  renameSync(temporary,file);
}
export function immutableEnvelope(file, value) {
  const bytes=JSON.stringify(value)+'\n';
  if(existsSync(file)) {
    if(readFileSync(file,'utf8')!==bytes)throw protocolError('IMMUTABLE_ENVELOPE_CONFLICT');
    return file;
  }
  const temporary=`${file}.${process.pid}.${randomUUID()}.tmp`;
  const fd=openSync(temporary,'wx',0o600);
  try {writeFileSync(fd,bytes);fsyncSync(fd);}finally{closeSync(fd);}
  try {linkSync(temporary,file);}catch(error) {
    if(error.code!=='EEXIST')throw error;
    if(readFileSync(file,'utf8')!==bytes)throw protocolError('IMMUTABLE_ENVELOPE_CONFLICT');
  }finally{unlinkSync(temporary);}
  return file;
}
const alive = pid => {try {process.kill(pid,0);return true;}catch(error){return error.code!=='ESRCH';}};
export function claimLane(root,lane) {
  const parent=path.join(root,'sync-locks');mkdirSync(parent,{recursive:true,mode:0o700});
  const directory=path.join(parent,hash(lane));
  for(let retry=0;retry<2;retry++) {
    try {mkdirSync(directory,{mode:0o700});}
    catch(error) {
      if(error.code!=='EEXIST') throw error;
      let owners;
      try {owners=readdirSync(directory);}catch {return null;}
      if(!owners.length) {if(Date.now()-statSync(directory).mtimeMs<60000)return null;}
      else {
        if(owners.some(name=>!/^\d+-[a-f0-9-]+\.owner$/.test(name) || alive(Number(name.split('-')[0])))) return null;
        for(const name of owners) try {unlinkSync(path.join(directory,name));}catch(error){if(error.code!=='ENOENT')return null;}
      }
      try {rmdirSync(directory);}catch {return null;}
      continue;
    }
    const owner=path.join(directory,`${process.pid}-${randomUUID()}.owner`);
    writeFileSync(owner,'',{flag:'wx',mode:0o600});
    return ()=>{try {unlinkSync(owner);rmdirSync(directory);}catch(error){if(error.code!=='ENOENT')throw error;}};
  }
  return null;
}
export function queueItems(directory,kind) {
  if(!existsSync(directory))return [];
  return readdirSync(directory).filter(name=>/^[a-zA-Z0-9_.:-]+\.json$/.test(name)).map(name=>{
    const filePath=path.join(directory,name);
    try {
      const raw=readFileSync(filePath), value=JSON.parse(raw);
      const item={filePath,kind,...(kind==='event'?{payload:value}:value),envelope_hash:hash(raw)};
      const payload=kind==='event'?item.payload?.event:item.payload;
      item.lane=typeof payload?.session_id==='string' && payload.session_id ? `session:${payload.session_id}`:'legacy';
      item.time=payload?.captured_at || payload?.occurred_at || '';
      if(typeof item.time!=='string') {item.time='';item.parse_error=true;}
      item.phase=payload?.phase;
      delete item.payload;
      Object.defineProperty(item,'payload',{enumerable:true,get(){const value=JSON.parse(readFileSync(this.filePath,'utf8'));return this.kind==='event'?value:value.payload;}});
      return item;
    }catch(error){return {filePath,kind,lane:'legacy',time:'',parse_error:true};}
  }).sort((a,b)=>a.time.localeCompare(b.time) || (['acknowledged','failed'].includes(a.phase)?1:0)-(['acknowledged','failed'].includes(b.phase)?1:0) || a.filePath.localeCompare(b.filePath));
}
export function queueState(item) {
  try {
    if(!existsSync(item.filePath+'.state')) return {state:'pending',attempt_count:0,lane:item.lane};
    const value=JSON.parse(readFileSync(item.filePath+'.state','utf8'));
    if(!states.has(value.state) || !Number.isInteger(value.attempt_count) || value.attempt_count<0
      || (value.last_error_code && !/^[A-Z0-9_]{1,64}$/.test(value.last_error_code))
      || (value.next_retry_at && !Number.isFinite(Date.parse(value.next_retry_at)))) throw Error();
    return value;
  } catch {return {state:'blocked_protocol',last_error_code:'CORRUPT_SIDECAR',attempt_count:0,lane:item.lane,corrupt:true};}
}
export function queueSummary(items,now=Date.now(),root=null) {
  const counts=Object.fromEntries([...states].map(state=>[state,0]));
  let oldest=null,last=null;
  for(const item of items) {
    const state=queueState(item); counts[item.parse_error?'blocked_protocol':state.state]++;
    const time=Date.parse(item.time); if(Number.isFinite(time))oldest=Math.max(oldest??0,now-time);
    if(state.last_error_code)last=state.last_error_code;
  }
  let lastSuccess=null;
  try {if(root)lastSuccess=JSON.parse(readFileSync(path.join(root,'sync-last-success.state'),'utf8')).at;}catch {}
  return {counts,queued:items.length,oldest_age_ms:oldest,high_water:items.length>=1000,last_error_code:last,last_success_at:lastSuccess};
}
export async function flushQueue(items, {root,credential,send,accepted=()=>{},quarantine=null,predecessors=[],now=()=>Date.now(),random=Math.random,afterAccepted=()=>{},maxElapsedMs=15000}) {
  const credentialId=hash(credential), blocked=new Set(), result={queued_before:items.length,flushed:0,quarantined:0,blocked:0};
  const markerDir=path.join(root,'sync-lanes');mkdirSync(markerDir,{recursive:true,mode:0o700});
  const authFile=path.join(root,'sync-auth.state');
  try {if(existsSync(authFile)&&JSON.parse(readFileSync(authFile,'utf8')).credential_fingerprint===credentialId)return {...result,blocked:items.length};}catch{return {...result,blocked:items.length};}
  let attempts=0;const started=performance.now();
  for(const item of items) {
    if(attempts>=100 || performance.now()-started>=maxElapsedMs)break;
    const release=claimLane(root,item.lane); if(!release){result.blocked++;continue;}
    try {
      if(!existsSync(item.filePath))continue;
      const markerPath=path.join(markerDir,hash(item.lane)+'.state');
      let marker=null;
      try {if(existsSync(markerPath))marker=JSON.parse(readFileSync(markerPath,'utf8'));}catch {result.blocked++;continue;}
      let state=queueState(item);
      const fail=error=>{
        state=retryState(error,state,{now:now(),random,credential:credentialId});
        state.lane=item.lane;state.envelope_hash=item.envelope_hash;
        atomicState(item.filePath+'.state',state);
        atomicState(markerPath,{file:item.filePath,state:state.state});
        if(state.state==='paused_auth')atomicState(authFile,{credential_fingerprint:credentialId,last_error_code:state.last_error_code});
        if(state.state==='quarantined' && quarantine)quarantine(item,error);
        blocked.add(item.lane);result[state.state==='quarantined'?'quarantined':'blocked']++;
      };
      if(state.corrupt) {atomicState(markerPath,{file:item.filePath,state:'blocked_protocol'});result.blocked++;continue;}
      if(predecessors.some(other=>other.filePath!==item.filePath && other.lane===item.lane && other.time<item.time && existsSync(other.filePath))) {
        atomicState(item.filePath+'.state',{...state,state:'blocked_gap',last_error_code:'PENDING_PREDECESSOR'});result.blocked++;continue;
      }
      if(blocked.has(item.lane) || (marker && marker.file!==item.filePath)) {
        if(!['quarantined','blocked_protocol','blocked_reconciliation'].includes(state.state))atomicState(item.filePath+'.state',{...state,state:'blocked_gap',last_error_code:'LANE_GAP',lane:item.lane});
        result.blocked++;continue;
      }
      if(item.parse_error){fail(protocolError('CORRUPT_ENVELOPE',422));continue;}
      if(hash(readFileSync(item.filePath))!==item.envelope_hash){fail(protocolError('RECEIPT_MISMATCH'));continue;}
      if(state.envelope_hash && state.envelope_hash!==item.envelope_hash){fail(protocolError('RECEIPT_MISMATCH'));continue;}
      if(['quarantined','blocked_protocol','blocked_reconciliation'].includes(state.state)
        || (state.state==='paused_auth' && state.credential_fingerprint===credentialId)
        || (state.credential_fingerprint===credentialId && state.next_retry_at && Date.parse(state.next_retry_at)>now())) {blocked.add(item.lane);result.blocked++;continue;}
      try {
        attempts++;
        atomicState(item.filePath+'.state',{...state,state:'pending',envelope_hash:item.envelope_hash,attempt_count:state.attempt_count+1,last_attempt_at:new Date(now()).toISOString()});
        let response;
        try {response=await send(item);}catch(error){
          if(error.statusCode!==409 || error.responseData?.error_code!=='IDEMPOTENT_REPLAY')throw error;
          response=error.responseData;
        }
        validateAcceptance(item.kind,item,response);
        await accepted(item,response);
        Object.defineProperty(result,'last_response',{value:response,configurable:true});
        await afterAccepted(item);
        atomicState(path.join(root,'sync-last-success.state'),{at:new Date(now()).toISOString()});
        if(existsSync(authFile))unlinkSync(authFile);
        unlinkSync(item.filePath);
        for(const file of [item.filePath+'.state',markerPath])if(existsSync(file))unlinkSync(file);
        result.flushed++;
      } catch(error){fail(error);if(state.state==='paused_auth')break;}
    } finally {release();}
  }
  return result;
}

File: adapters/hermes/sync_protocol.py
"""Stdlib transport and durable retry contract shared semantically with the JS adapters."""
import hashlib
import http.client
import json
import math
import os
import queue as result_queue
import random
import re
import socket
import threading
import time
import urllib.error
import urllib.request
import uuid
from datetime import datetime, timezone
from email.utils import parsedate_to_datetime
from pathlib import Path

RESPONSE_LIMIT = 2 * 1024 * 1024
_DNS_SLOTS = threading.BoundedSemaphore(8)

def failure(code, status=None, retry_after=None):
    error = RuntimeError(f"Mnemuron transport: {code}")
    error.error_code, error.status_code, error.retry_after = code, status, retry_after
    return error

class _DeadlineConnection:
    def _connect_resolved(self, address, timeout, source_address=None):
        # OS name resolution has no cancellable timeout. Bound its wait and the
        # number of outstanding daemon workers; late answers never open sockets.
        if not _DNS_SLOTS.acquire(blocking=False): raise failure('RESOLVER_BUSY')
        results = result_queue.Queue(maxsize=1)
        def resolve():
            try: results.put((socket.getaddrinfo(*address, 0, socket.SOCK_STREAM), None))
            except Exception as error: results.put((None, error))
            finally: _DNS_SLOTS.release()
        threading.Thread(target=resolve,daemon=True).start()
        try: addresses, error = results.get(timeout=max(.001,self._deadline-time.monotonic()))
        except result_queue.Empty: raise failure('TOTAL_TIMEOUT')
        if error: raise error
        last = None
        for family, kind, protocol, _, target in addresses:
            remaining=self._deadline-time.monotonic()
            if remaining <= 0 or self._expired: raise failure('TOTAL_TIMEOUT')
            sock=socket.socket(family,kind,protocol);self._deadline_socket=sock
            try:
                sock.settimeout(remaining)
                if source_address: sock.bind(source_address)
                sock.connect(target)
                remaining=self._deadline-time.monotonic()
                if remaining<=0: raise failure('TOTAL_TIMEOUT')
                sock.settimeout(remaining)
                return sock
            except Exception as error:
                sock.close();last=error
        if last: raise last
        raise failure('NETWORK_ERROR')

    def getresponse(self):
        try:
            response = super().getresponse()
            response._mnemuron_timer = self._timer
            response._mnemuron_expired = lambda: self._expired
            return response
        except Exception:
            if hasattr(self, '_timer'): self._timer.cancel()
            raise

    def connect(self):
        self._expired = False
        self._deadline_socket = None
        self._deadline = time.monotonic()+self.timeout
        self._create_connection = self._connect_resolved
        def expire():
            self._expired = True
            sock = self.sock or self._deadline_socket
            if sock:
                try: sock.shutdown(socket.SHUT_RDWR)
                except OSError: pass
        self._timer = threading.Timer(self.timeout, expire)
        self._timer.daemon = True
        self._timer.start()
        super().connect()
        self._deadline_socket = self.sock
        if self._expired:
            self.sock.close()
            raise failure("TOTAL_TIMEOUT")

class _HTTPConnection(_DeadlineConnection, http.client.HTTPConnection): pass
class _HTTPSConnection(_DeadlineConnection, http.client.HTTPSConnection): pass
class _HTTPHandler(urllib.request.HTTPHandler):
    def http_open(self, request): return self.do_open(_HTTPConnection, request)
class _HTTPSHandler(urllib.request.HTTPSHandler):
    def https_open(self, request): return self.do_open(_HTTPSConnection, request, context=self._context)
class _NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, *args, **kwargs): return None

def request_json(request, timeout):
    opener = urllib.request.build_opener(_HTTPHandler(), _HTTPSHandler(), _NoRedirect())
    try:
        response = opener.open(request, timeout=timeout)
    except urllib.error.HTTPError as error:
        response = error
    except (OSError, urllib.error.URLError, http.client.HTTPException) as error:
        raise failure('NETWORK_ERROR') from error
    try:
        with response:
            content = response.read(RESPONSE_LIMIT + 1)
            if getattr(response, '_mnemuron_expired', lambda: False)(): raise failure('TOTAL_TIMEOUT')
            if len(content) > RESPONSE_LIMIT: raise failure("RESPONSE_TOO_LARGE", response.code)
            try: data = json.loads(content.decode("utf-8"))
            except (ValueError, UnicodeError): data = None
            status = response.code
            if not 200 <= status < 300:
                code = data.get("error_code") if isinstance(data, dict) else None
                error = failure(code if isinstance(code, str) and re.fullmatch(r'[A-Z][A-Z0-9_]{0,63}', code) else ("REDIRECT_BLOCKED" if 300 <= status < 400 else f"HTTP_{status}"), status, response.headers.get("Retry-After"))
                error.response_data = data
                raise error
            if not isinstance(data, dict): raise failure("INVALID_RESPONSE_JSON", status)
            return data
    except (TimeoutError, OSError, http.client.HTTPException) as error:
        raise failure("NETWORK_ERROR") from error
    finally:
        timer = getattr(response, '_mnemuron_timer', None)
        if timer: timer.cancel()

def retry_state(error, previous=None, *, now=None, rng=random.random, credential=""):
    now = time.time() * 1000 if now is None else now
    previous = previous or {}
    status = getattr(error, "status_code", None)
    code = getattr(error, "error_code", "NETWORK_ERROR")
    count = previous.get("attempt_count", 0) + 1
    iso = lambda value: datetime.fromtimestamp(value / 1000, timezone.utc).isoformat(timespec="milliseconds").replace("+00:00", "Z")
    state = {**previous, "attempt_count": count, "last_error_code": code, "last_http_status": status,
             "first_failed_at": previous.get("first_failed_at") or iso(now), "next_retry_at": None, "credential_fingerprint": credential}
    if status in (401, 403): state["state"] = "paused_auth"
    elif status == 409 or code == "RECEIPT_MISMATCH": state["state"] = "blocked_reconciliation"
    elif status == 413 or (status in (400, 422) and code in ("INVALID_PAYLOAD", "CONTENT_TOO_LONG", "SCOPE_MISMATCH", "INVALID_MEMORY_SCOPE", "REQUEST_BODY_TOO_LARGE")): state["state"] = "quarantined"
    elif code in ("RESPONSE_TOO_LARGE", "INVALID_RESPONSE_JSON", "REDIRECT_BLOCKED") or (status and status < 500 and status != 429): state["state"] = "blocked_protocol"
    else:
        state["state"] = "retry_wait"
        delay = min(300000, 1000 * 2 ** min(count - 1, 18)) * (0.5 + rng() * 0.5)
        if status == 429:
            value = getattr(error, "retry_after", None)
            try: until = now + float(value) * 1000
            except (ValueError, TypeError):
                try: until = parsedate_to_datetime(value).timestamp() * 1000
                except (ValueError, TypeError, OverflowError): until = now
            if math.isfinite(until) and now < until < 253402300799000: delay = max(delay, until - now)
        state["next_retry_at"] = iso(now + delay)
    return state

def atomic(file, value):
    temporary = file.with_name(file.name + "." + str(uuid.uuid4()) + ".tmp")
    descriptor = os.open(temporary, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    with os.fdopen(descriptor, "w") as stream:
        json.dump(value, stream)
        stream.flush()
        os.fsync(stream.fileno())
    os.replace(temporary, file)

def immutable_envelope(file, value):
    file.parent.mkdir(parents=True, exist_ok=True, mode=0o700)
    content = (json.dumps(value, ensure_ascii=False, sort_keys=True) + '\n').encode()
    if file.exists():
        if json.loads(file.read_text()) != value: raise failure('IMMUTABLE_ENVELOPE_CONFLICT')
        return
    temporary = file.with_name(file.name + '.' + str(uuid.uuid4()) + '.tmp')
    descriptor = os.open(temporary, os.O_WRONLY | os.O_CREAT | os.O_EXCL, 0o600)
    with os.fdopen(descriptor, 'wb') as stream:
        stream.write(content)
        stream.flush()
        os.fsync(stream.fileno())
    try:
        try: os.link(temporary, file)
        except FileExistsError:
            if json.loads(file.read_text()) != value: raise failure('IMMUTABLE_ENVELOPE_CONFLICT')
    finally: temporary.unlink()

def queue_items(directory, kind):
    if not directory.exists(): return []
    items = []
    for file in sorted(directory.glob("*.json")):
        try:
            raw = file.read_bytes()
            value = json.loads(raw)
            item = {"file": file, "kind": kind, **({"payload": value} if kind == "event" else value), "envelope_hash": hashlib.sha256(raw).hexdigest()}
            payload = item["payload"].get("event", {}) if kind == "event" else item["payload"]
            item["lane"] = "session:" + payload["session_id"] if payload.get("session_id") else "legacy"
            item["time"] = payload.get("captured_at") or payload.get("occurred_at") or ""
            if not isinstance(item['time'],str): item['time']='';item['parse_error']=True
            item['phase'] = payload.get('phase')
            item.pop('payload')
        except (ValueError, KeyError, TypeError, AttributeError): item = {"file": file, "kind": kind, "lane": "legacy", "time": "", "parse_error": True}
        items.append(item)
    return sorted(items, key=lambda item: (item["time"], item.get('phase') in ("acknowledged", "failed"), str(item["file"])))

def validate_acceptance(kind, item, result):
    payload = item["payload"]
    if kind == "event":
        event_id = payload.get("event", {}).get("event_id")
        if isinstance(result, dict) and isinstance(event_id, str) and event_id and result.get("status") == "accepted":
            counts = (result.get("inserted"), result.get("duplicate"))
            if (type(result.get("received")) in (int, float) and result["received"] == 1
                and all(type(n) in (int, float) and n in (0, 1) for n in counts) and sum(counts) == 1
                and isinstance(result.get("accepted_event_ids"), list) and result["accepted_event_ids"] == [event_id]): return
    else:
        delivery = result.get("delivery", {})
        attempt = next((x for x in delivery.get("attempts", []) if x.get("attempt_id") == payload.get("attempt_id")), {})
        field = {"injected": "injected_at", "acknowledged": "acknowledged_at", "failed": "failed_at"}.get(payload.get("phase"))
        if (result.get("event_id") == payload.get("event_id") and result.get("inserted", -2) + result.get("duplicate", -2) == 1
            and delivery.get("resume_id") == item.get("resume_id") and delivery.get("preview_version") == payload.get("preview_version")
            and all(attempt.get(key) == payload.get(key) for key in ("session_id", "turn_id", "workstream_id"))
            and attempt.get(field) == payload.get("occurred_at") and payload.get("event_id") in attempt.get("event_ids", [])
            and (payload.get("phase") != "acknowledged" or attempt.get("ack_complete") is True)): return
    raise failure("RECEIPT_MISMATCH")

def flush_queue(items, *, root, credential, send, predecessors=(), now=lambda: time.time() * 1000, rng=random.random, max_elapsed_ms=15000):
    # flock is process-owned, released by the kernel on death, and never stolen by mtime.
    import fcntl
    credential = hashlib.sha256(credential.encode()).hexdigest()
    markers = root / "sync-lanes"
    markers.mkdir(parents=True, exist_ok=True, mode=0o700)
    result = {"queued_before": len(items), "flushed": 0, "quarantined": 0, "blocked": 0}
    auth_file = root / 'sync-auth.state'
    try:
        if auth_file.exists() and json.loads(auth_file.read_text()).get('credential_fingerprint') == credential:
            return {**result, 'blocked': len(items)}
    except (ValueError, AttributeError): return {**result, 'blocked': len(items)}
    attempts = 0
    started = time.monotonic()
    for item in items:
        if attempts >= 100 or (time.monotonic()-started)*1000 >= max_elapsed_ms: break
        key = hashlib.sha256(item["lane"].encode()).hexdigest()
        lockfile = markers / (key + ".lock")
        with os.fdopen(os.open(lockfile, os.O_CREAT | os.O_RDWR, 0o600), "w") as lock:
            try: fcntl.flock(lock, fcntl.LOCK_EX | fcntl.LOCK_NB)
            except BlockingIOError:
                result["blocked"] += 1
                continue
            if not item["file"].exists(): continue
            sidecar = Path(str(item["file"]) + ".state")
            marker = markers / (key + ".state")
            try:
                state = json.loads(sidecar.read_text()) if sidecar.exists() else {"state": "pending", "attempt_count": 0, "lane": item["lane"]}
                owner = json.loads(marker.read_text()).get("file") if marker.exists() else None
                if (not isinstance(state.get("attempt_count"), int) or state['attempt_count'] < 0
                    or (state.get('last_error_code') and not re.fullmatch(r'[A-Z0-9_]{1,64}',str(state['last_error_code'])))
                    or state.get('state') not in ('pending','retry_wait','paused_auth','quarantined','blocked_protocol','blocked_reconciliation','blocked_gap')): raise ValueError()
                if state.get('next_retry_at'): datetime.fromisoformat(state['next_retry_at'].replace('Z', '+00:00'))
                if any(other['lane'] == item['lane'] and other['time'] < item['time'] and other['file'].exists() for other in predecessors):
                    atomic(sidecar, {**state, 'state':'blocked_gap', 'last_error_code':'PENDING_PREDECESSOR'})
                    result['blocked'] += 1
                    continue
                if owner and owner != str(item["file"]):
                    atomic(sidecar, {**state, "state": "blocked_gap", "last_error_code": "LANE_GAP"})
                    result["blocked"] += 1
                    continue
            except (ValueError, AttributeError):
                atomic(marker, {"file": str(item["file"]), "state": "blocked_protocol"})
                result["blocked"] += 1
                continue
            if state["state"] in ("quarantined", "blocked_protocol", "blocked_reconciliation") or (state["state"] == "paused_auth" and state.get("credential_fingerprint") == credential):
                result["blocked"] += 1
                continue
            try:
                if state.get('credential_fingerprint') == credential and state.get("next_retry_at") and datetime.fromisoformat(state["next_retry_at"].replace("Z", "+00:00")).timestamp()*1000 > now():
                    result["blocked"] += 1
                    continue
                if item.get("parse_error"): raise failure("CORRUPT_ENVELOPE", 422)
                if state.get("envelope_hash") and state["envelope_hash"] != item["envelope_hash"]: raise failure("RECEIPT_MISMATCH")
                raw = item['file'].read_bytes()
                if hashlib.sha256(raw).hexdigest() != item['envelope_hash']: raise failure('RECEIPT_MISMATCH')
                value = json.loads(raw)
                item['payload'] = value if item['kind'] == 'event' else value['payload']
                attempts += 1
                atomic(sidecar, {**state,'state':'pending','attempt_count':state['attempt_count']+1,'envelope_hash':item['envelope_hash']})
                try: response = send(item)
                except Exception as error:
                    data = getattr(error, "response_data", None)
                    if getattr(error, "status_code", None) != 409 or not isinstance(data, dict) or data.get("error_code") != "IDEMPOTENT_REPLAY": raise
                    response = data
                validate_acceptance(item["kind"], item, response)
                atomic(root / 'sync-last-success.state', {'at':datetime.fromtimestamp(now()/1000, timezone.utc).isoformat()})
                auth_file.unlink(missing_ok=True)
                item["file"].unlink()
                sidecar.unlink(missing_ok=True)
                marker.unlink(missing_ok=True)
                result["flushed"] += 1
            except Exception as error:
                state = retry_state(error, state, now=now(), rng=rng, credential=credential)
                state.update(lane=item["lane"], envelope_hash=item.get("envelope_hash"))
                atomic(sidecar, state)
                atomic(marker, {"file": str(item["file"]), "state": state["state"]})
                result["quarantined" if state["state"] == "quarantined" else "blocked"] += 1
                if state['state'] == 'paused_auth':
                    atomic(auth_file, {'credential_fingerprint':credential,'last_error_code':state['last_error_code']})
                    break
            finally: item.pop('payload', None)
    return result

def queue_summary(items, root=None):
    counts = {}
    oldest = None
    last = None
    for item in items:
        try:
            file = Path(str(item["file"]) + ".state")
            data = json.loads(file.read_text()) if file.exists() else {'state':'pending','attempt_count':0}
            state = data.get('state')
            if (state not in ('pending','retry_wait','paused_auth','quarantined','blocked_protocol','blocked_reconciliation','blocked_gap')
                or not isinstance(data.get('attempt_count'),int)
                or (data.get('last_error_code') and not re.fullmatch(r'[A-Z0-9_]{1,64}',str(data['last_error_code'])))):
                raise ValueError()
            last = data.get('last_error_code') or last
            if item.get("parse_error"): state = "blocked_protocol"
        except (ValueError, AttributeError): state = "blocked_protocol";last='CORRUPT_SIDECAR'
        counts[state] = counts.get(state, 0) + 1
        try:
            age = time.time()*1000 - datetime.fromisoformat(item['time'].replace('Z','+00:00')).timestamp()*1000
            oldest = max(oldest or 0, age)
        except (ValueError, KeyError): pass
    success = None
    try:
        if root: success = json.loads((root / 'sync-last-success.state').read_text()).get('at')
    except (OSError, ValueError, AttributeError): pass
    return {"counts": counts, "queued": len(items), "high_water": len(items) >= 1000, 'oldest_age_ms':oldest,'last_error_code':last,'last_success_at':success}
